### DISCUSSION TITLE
Sort based on file attributes?

### DISCUSSION THREAD

**Discussion by shrik450** [minad/vertico#482] [Q&A] [answered]
Hey, I'm trying to write a custom sort function for vertico that sorts files based on the attributes returned by `directory-files-and-attributes`. I would expect the candidates sequence that is given to my sort function to be the lists that `directory-files-and-attributes` returns. However, I'm seeing vertico convert this into just the filename before returning passing them onto the function, making it impossible for me to use those attributes to sort. Is there a setting I can disable to make this work?

**Comment by minad** [comment, marked as answer]
Vertico (and generally completion UIs) operate on string, so there is no direct way to make this work. But of course you can look up the attributes again given the file name in your sort function.

**Reply by shrik450** [reply]
I understand that the completion can only work on strings, but is there a reason why the *sort function* has to be called on strings? To put it another way, what would be the problem with converting to strings *after* sorting?

Either way, getting the file attributes for every file when sorting is untenable. The folders I'm working with have hundreds of, if not over a thousand files. I've worked around this for now by converting the arguments to strings myself up front and including the attributes I want in the string. My sort function then splits the strings, sorts based on the attributes and returns only the part of the string I care about.

**Reply by minad** [reply]
> I understand that the completion can only work on strings, but is there a reason why the sort function has to be called on strings?

That's the way it is defined in the Emacs completion system. See for example display-sort-function and cycle-sort-function.
